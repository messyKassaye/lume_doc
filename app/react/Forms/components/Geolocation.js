import PropTypes from 'prop-types';
import React, { Component } from 'react';
import Map from 'app/Map/Map';
import { Translate } from 'app/I18N';

const defaultValue = { lat: '', lon: '', label: '' };

function isCoordinateValid(coord) {
  return typeof coord === 'number' && !Number.isNaN(coord);
}

export default class Geolocation extends Component {
  constructor(props) {
    super(props);
    this.latChange = this.latChange.bind(this);
    this.lonChange = this.lonChange.bind(this);
    this.mapClick = this.mapClick.bind(this);
    this.clearCoordinates = this.clearCoordinates.bind(this);
    this.emptyValue = [];
    const { lat, lon } = this.getInputValues();

    this.state = {
      currentLatitude: lat,
      currentLongitude: lon,
    };
  }

  onChange(newValue) {
    this.setState({ currentLatitude: newValue.lat, currentLongitude: newValue.lon });
    const { onChange, value } = this.props;

    if (!isCoordinateValid(newValue.lat) || !isCoordinateValid(newValue.lon)) {
      onChange(this.emptyValue);
      return;
    }
    const valueToSend = value.slice(1);
    valueToSend.unshift(newValue);
    onChange(valueToSend);
  }

  getInputValues() {
    const { value } = this.props;
    const { lat, lon, label } = value && value[0] ? value[0] : { ...defaultValue };
    return { lat, lon, label };
  }

  latChange(e) {
    let latitude = e.target.value ? parseFloat(e.target.value) : '';
    latitude = latitude && latitude < -89.99999 ? -89.99999 : latitude;
    latitude = latitude && latitude > 90 ? 90 : latitude;

    const { label } = this.getInputValues();
    const { currentLongitude } = this.state;
    this.onChange({ lat: latitude, lon: currentLongitude, label });
  }

  lonChange(e) {
    const longitude = e.target.value ? parseFloat(e.target.value) : '';

    const { label } = this.getInputValues();
    const { currentLatitude } = this.state;
    this.onChange({ lat: currentLatitude, lon: longitude, label });
  }

  mapClick(event) {
    const { label } = this.getInputValues();
    this.onChange({ lat: parseFloat(event.lngLat[1]), lon: parseFloat(event.lngLat[0]), label });
  }

  clearCoordinates() {
    this.setState({ currentLatitude: '', currentLongitude: '' });
    const { onChange } = this.props;
    onChange(this.emptyValue);
  }

  render() {
    const { currentLatitude, currentLongitude } = this.state;
    const markers = [];

    if (isCoordinateValid(currentLatitude) && isCoordinateValid(currentLongitude)) {
      markers.push({
        latitude: parseFloat(currentLatitude),
        longitude: parseFloat(currentLongitude),
      });
    }

    return (
      <div className="geolocation form-inline">
        <Map markers={markers} onClick={this.mapClick} height={370} autoCenter={false} />
        <div className="form-row">
          <div className="form-group half-width">
            <label>Latitude</label>
            <input
              onChange={this.latChange}
              className="form-control"
              type="number"
              id="lat"
              value={currentLatitude}
              step="any"
            />
          </div>
          <div className="form-group half-width">
            <label>Longitude</label>
            <input
              onChange={this.lonChange}
              className="form-control"
              type="number"
              id="lon"
              value={currentLongitude}
              step="any"
            />
          </div>
        </div>
        {(currentLatitude || currentLongitude) && (
          <div className="clear-field-button">
            <button type="button" onClick={this.clearCoordinates}>
              <Translate>Clear coordinates</Translate>
            </button>
          </div>
        )}
      </div>
    );
  }
}

Geolocation.defaultProps = {
  value: [{ ...defaultValue }],
};

Geolocation.propTypes = {
  value: PropTypes.arrayOf(PropTypes.object),
  onChange: PropTypes.func.isRequired,
};
