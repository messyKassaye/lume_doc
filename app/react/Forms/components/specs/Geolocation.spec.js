import React from 'react';
import { shallow } from 'enzyme';

import Geolocation from '../Geolocation';

describe('Geolocation', () => {
  let component;
  let instance;
  let props;

  const render = () => {
    component = shallow(<Geolocation {...props} />);
    instance = component.instance();
  };

  beforeEach(() => {
    props = {
      value: [
        { lat: 32.18, lon: -17.2, label: 'home' },
        { lat: 13.07, lon: 5.1, label: 'Created through migration?' },
      ],
      onChange: jasmine.createSpy('onChange'),
    };

    render();
  });

  it('should render 2 inputs with the lat and lon values, and optional label', () => {
    const inputs = component.find('input');
    const latInput = inputs.at(0);
    const lonInput = inputs.at(1);

    expect(latInput.props().value).toBe(32.18);
    expect(lonInput.props().value).toBe(-17.2);
  });

  describe('when lat changes', () => {
    it('with invalid coordinated should call onChange with empty value', () => {
      const latInput = component.find('input').at(0);
      latInput.simulate('change', { target: { value: '' } });
      expect(props.onChange).toHaveBeenCalledWith([]);
    });

    function expectOnChangeCallWhenInputSimulation(simulatedInput, expectedValue) {
      const latInput = component.find('input').at(0);
      latInput.simulate('change', { target: { value: simulatedInput } });
      expect(props.onChange).toHaveBeenCalledWith([
        { lat: expectedValue, lon: -17.2, label: 'home' },
        props.value[1],
      ]);
    }

    it('should call onChange with the new value', () => {
      expectOnChangeCallWhenInputSimulation('0', 0);
      expectOnChangeCallWhenInputSimulation('1', 1);
      expectOnChangeCallWhenInputSimulation('19', 19);
    });

    it('with an angle lower than -89.99999 degrees the value should be replace by -89.99999 degrees', () => {
      expectOnChangeCallWhenInputSimulation('-91', -89.99999);
      expectOnChangeCallWhenInputSimulation('-120.34', -89.99999);
      expectOnChangeCallWhenInputSimulation('-90', -89.99999);
      expectOnChangeCallWhenInputSimulation('-89.999991', -89.99999);
    });

    it('with an angle greater than 90 degrees the value should be replace by 90 degrees', () => {
      expectOnChangeCallWhenInputSimulation('91', 90);
      expectOnChangeCallWhenInputSimulation('120.34', 90);
    });
  });

  describe('when lon changes', () => {
    let lonInput;

    beforeEach(() => {
      lonInput = component.find('input').at(1);
    });

    it('should call onChange with the new value', () => {
      lonInput.simulate('change', { target: { value: '28' } });
      expect(props.onChange).toHaveBeenCalledWith([
        { lat: 32.18, lon: 28, label: 'home' },
        props.value[1],
      ]);
    });

    it('should call onChange with empty value when invalid longitude', () => {
      lonInput.simulate('change', { target: { value: '' } });
      expect(props.onChange).toHaveBeenCalledWith([]);
    });
  });

  describe('mapClick', () => {
    it('should call onChange with the map value', () => {
      const event = { lngLat: [5, 13] };
      instance.mapClick(event);
      expect(props.onChange).toHaveBeenCalledWith([
        { lat: 13, lon: 5, label: 'home' },
        props.value[1],
      ]);
      expect(
        component
          .find('input')
          .at(0)
          .props().value
      ).toEqual(13);
      expect(
        component
          .find('input')
          .at(1)
          .props().value
      ).toEqual(5);
    });

    it('should work assign default values if original point was null', () => {
      props.value = [null];
      render();
      instance.mapClick({ lngLat: [13, 7] });
      expect(props.onChange).toHaveBeenCalledWith([{ lat: 7, lon: 13, label: '' }]);
    });
  });

  describe('empty lat/lon values', () => {
    function testInputWillTriggerOnChangeWithoutValue(getInput) {
      const inputs = component.find('input');
      const input = getInput(inputs);
      input.simulate('change', { target: { value: '' } });
      expect(props.onChange).toHaveBeenCalledWith([]);
    }

    describe('if lon is empty and lat is set to empty', () => {
      it('should call onChange without a value', () => {
        props.value[0].lon = '';
        render();
        testInputWillTriggerOnChangeWithoutValue(inputs => inputs.at(0));
      });
    });

    describe('if lat is empty and lon is set to empty', () => {
      it('should call onChange without a value', () => {
        props.value[0].lat = '';
        render();
        testInputWillTriggerOnChangeWithoutValue(inputs => inputs.at(1));
      });
    });

    describe('when they are no empty anymore', () => {
      it('should call onChange with the correct values', () => {
        component
          .find('input')
          .at(0)
          .simulate('change', { target: { value: '' } });
        component
          .find('input')
          .at(1)
          .simulate('change', { target: { value: '' } });

        component
          .find('input')
          .at(0)
          .simulate('change', { target: { value: '1' } });
        component
          .find('input')
          .at(1)
          .simulate('change', { target: { value: '2' } });

        expect(props.onChange).toHaveBeenCalledWith([
          { lat: 1, lon: 2, label: 'home' },
          { label: 'Created through migration?', lat: 13.07, lon: 5.1 },
        ]);
      });
    });
  });

  describe('should render button to clear fields', () => {
    function expectRenderButton(latitude, longitude) {
      component
        .find('input')
        .at(0)
        .simulate('change', { target: { value: latitude } });
      component
        .find('input')
        .at(1)
        .simulate('change', { target: { value: longitude } });
      expect(component.exists('.clear-field-button')).toEqual(true);
    }

    it('when latitude empty and longitude not empty', () => {
      expectRenderButton('', '1');
    });

    it('when longitude empty and latitude not empty', () => {
      expectRenderButton('1', '');
    });

    it('when longitude and latitude not empty', () => {
      expect(component).toMatchSnapshot();
    });
  });

  describe('should hide clear fields button ', () => {
    it('when latitude and longitude are empty', () => {
      props.value = [{ lat: '', lon: '' }];
      render();
      expect(component).toMatchSnapshot();
    });
  });

  describe('when clear fields button is clicked', () => {
    beforeEach(() => {
      const button = component.find('.clear-field-button button').first();
      button.simulate('click');
    });

    it('should call onChange without a value and remove the inputs values', () => {
      expect(props.onChange).toHaveBeenCalledWith([]);
    });

    it('should remove the inputs values', () => {
      expect(
        component
          .find('input')
          .at(0)
          .props().value
      ).toEqual('');
      expect(
        component
          .find('input')
          .at(1)
          .props().value
      ).toEqual('');
    });
  });
});
