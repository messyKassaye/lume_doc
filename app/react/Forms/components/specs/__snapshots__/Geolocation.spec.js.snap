// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Geolocation should hide clear fields button  when latitude and longitude are empty 1`] = `
<div
  className="geolocation form-inline"
>
  <Map
    autoCenter={false}
    clickOnCluster={[Function]}
    clickOnMarker={[Function]}
    cluster={false}
    height={370}
    hoverOnMarker={[Function]}
    interactiveLayerIds={Array []}
    latitude={null}
    longitude={null}
    markers={Array []}
    onClick={[Function]}
    renderMarker={null}
    renderPopupInfo={null}
    scrollZoom={true}
    showControls={false}
    width={null}
    zoom={4}
  />
  <div
    className="form-row"
  >
    <div
      className="form-group half-width"
    >
      <label>
        Latitude
      </label>
      <input
        className="form-control"
        id="lat"
        onChange={[Function]}
        step="any"
        type="number"
        value=""
      />
    </div>
    <div
      className="form-group half-width"
    >
      <label>
        Longitude
      </label>
      <input
        className="form-control"
        id="lon"
        onChange={[Function]}
        step="any"
        type="number"
        value=""
      />
    </div>
  </div>
</div>
`;

exports[`Geolocation should render button to clear fields when longitude and latitude not empty 1`] = `
<div
  className="geolocation form-inline"
>
  <Map
    autoCenter={false}
    clickOnCluster={[Function]}
    clickOnMarker={[Function]}
    cluster={false}
    height={370}
    hoverOnMarker={[Function]}
    interactiveLayerIds={Array []}
    latitude={null}
    longitude={null}
    markers={
      Array [
        Object {
          "latitude": 32.18,
          "longitude": -17.2,
        },
      ]
    }
    onClick={[Function]}
    renderMarker={null}
    renderPopupInfo={null}
    scrollZoom={true}
    showControls={false}
    width={null}
    zoom={4}
  />
  <div
    className="form-row"
  >
    <div
      className="form-group half-width"
    >
      <label>
        Latitude
      </label>
      <input
        className="form-control"
        id="lat"
        onChange={[Function]}
        step="any"
        type="number"
        value={32.18}
      />
    </div>
    <div
      className="form-group half-width"
    >
      <label>
        Longitude
      </label>
      <input
        className="form-control"
        id="lon"
        onChange={[Function]}
        step="any"
        type="number"
        value={-17.2}
      />
    </div>
  </div>
  <div
    className="clear-field-button"
  >
    <button
      onClick={[Function]}
      type="button"
    >
      <Connect(Translate)>
        Clear coordinates
      </Connect(Translate)>
    </button>
  </div>
</div>
`;
